src/Altinn.Apps/AppFrontend/react/altinn-app-frontend/src/components/base/CheckboxesContainerComponent.tsx

```diff
@@ -67,95 +68,46 @@ const useStyles = makeStyles({
   },
 });
 
-function usePrevious(value) {
-  const ref = React.useRef();
-  React.useEffect(() => {
-    ref.current = value.slice();
-  });
 
-  return ref.current;
-}
+const emptyList: undefined[] = []; // constant for reference stability
 
 export const CheckboxContainerComponent = (props: ICheckboxContainerProps) => {
   const classes = useStyles(props);
-  const apiOptions = useAppSelector(state => state.optionState.options[props.optionsId]);
-  const options = apiOptions || props.options || [];
-  const [selected, setSelected] = React.useState([]);
-  const prevSelected: any = usePrevious(selected);
+  const apiOptions: IOption[] = useAppSelector(state => state.optionState.options[props.optionsId]);
+  const options = apiOptions || props.options || emptyList;
   const checkBoxesIsRow: boolean = options.length <= 2;
 
-  React.useEffect(() => {
-    returnState();
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [options]);
+  const selected = React.useMemo(() => {
+    return props.formData.simpleBinding.split(',').map(v => options.find(o => o.value == v)).filter(o => !!o) ?? emptyList;
+  }, [props.formData, options])
 
+  // Implement preselected functionality
   React.useEffect(() => {
-    returnState();
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [props.formData?.simpleBinding]);
-
-  const returnState = () => {
+    const preselectedOption = options?.[props.preselectedOptionIndex]?.value;
     if (
-      !props.formData?.simpleBinding &&
-      props.preselectedOptionIndex >= 0 &&
-      options &&
-      props.preselectedOptionIndex < options.length
+      !selected &&
+      preselectedOption
     ) {
-      const preSelected: string[] = [];
-      preSelected[props.preselectedOptionIndex] =
-        options[props.preselectedOptionIndex].value;
-      props.handleDataChange(preSelected[props.preselectedOptionIndex]);
-      setSelected(preSelected);
-    } else {
-      setSelected(props.formData?.simpleBinding ? props.formData.simpleBinding.split(',') : []);
+      props.handleDataChange(preselectedOption);
     }
-  };
-
-  const onDataChanged = (event: React.ChangeEvent<HTMLInputElement>) => {
-    const newSelected: any = selected.slice();
-    const index = newSelected.indexOf(event.target.name);
+    // This effect should only run when new options are loaded
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
I think this should be removed. If we later are introducing other dependencies that are needed, this eslint-disable might cause a bug since we might forget to add it to the list.

It looks to me it doesnt matter if this is run several times either, because you will only call the `handleDataChange` if you have not selected anything, so removing it should be fine I think.

Reply:
Yes, I think `handleDataChange` should be set as a dependency. Ignoring it may lead to stale closures, which is very bad. Listing it may cause "unnescessary renders" because of the issue that its not referentially stable. If the unnescessary rerenders are causing performance issues, then that should be looked into, and the fix would probably be to wrap the `handleDataChange` in a `useCallback`. But if its not causing performance issues, I wouldnt worry about it.

Reply:
Stale closures is a non-issue with useEffect, because we create a fresh closure on every render, regardless of the deps array. The question is how we decide when the effect should be executed. This effect needs to run **only** when the list of options changes, so that we can select the default option if nothing was previously selected. If options did not change, there is no need to update the state. The easiest way to ensure that requirement is to ensure that options is referentially stable and put only options in the deps array.

Reply:
> Stale closures is a non-issue with useEffect, because we create a fresh closure on every render, regardless of the deps array

The execution of the hook will be skipped, unless any items in the dependency array changed (or if you have an empty array, it will only be run once. If you dont provide an array, it will be run on every render). [Stating that stale closures is a non-issue for useEffect is not correct. 
](https://reactjs.org/docs/hooks-faq.html#why-am-i-seeing-stale-props-or-state-inside-my-function)

This is why the exhaustive-deps rule exists -it is important to list all the dependencies, [even functions](https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies), so the useEffect runs when something changes, to avoid stale closures.

This effect should run also when `props.handleDataChange` is changed as well. Some other code might run and change that function to do something else, f.ex call another API. If this useEffect does not react to that change, it will still call the old `props.handleDataChange` function. The example is a bit contrived, and its highly unlikely that the handleDataChange will do something else at some point, but it illustrates the point.

Adding the comment to disable the exhaustive deps might be fine right now, because we know that `props.handleDataChange` will in fact not change. But it creates problems later, when some new code is introduced here, that uses other props that may change between renders. You might not notice the comment is there and the check is disabled so you forget to list the new dependency, and you introduce a stale closure and potentially a bug.

You already have a check to see if an option is selected in the useEffect, so it should not matter if the effect is executed on every render.

Reply:
No, the examples you link to sets callbacks in `useEffect` and obviously those callbacks needs to be updated in case they change, because they will be called at a much later time. The closure passed to `useEffect` here calls `handleDataChange` immediatly after render and there is no chance that reference to be stale.

The comment to disable the check now is fine because we **know** `handleDataChange` will change.

I'm tired of this discussion. Can you just add a suggestion to how you want the code to be written, so I can report the errors as bugs instead?

Reply:
> I'm tired of this discussion. Can you just add a suggestion to how you want the code to be written, so I can report the **behavioural** errors as bugs instead?

Whats with the attitude? I spend time on reviewing the code you wrote, and called out a problem - I think these types of replies are uncalled for. I just want to make sure we dont introduce new potential issues in the code, and not specifying the correct dependencies to the hooks array is a problem.

Either I am not understanding you correctly, or there is a misconception about how useEffect. But I think the docs are pretty clear on this; if you do not specify every external dependency in the dependency array, the outcome is unpredictable. It does not matter that it "calls `handleDataChange` immediately after a render", because the entire `useEffect` hook might be skipped in the next render, because it only cares if `options` changes. If `handleDataChange` is changed and `options` stay the same, the `useEffect` will not be run and the current reference to `handleDataChange` will be stale.

However, as I said earlier, this is not the main problem with the code, since `handleDataChange` is higly unliketly to be changed. The main problem is future maintainability. New dependencies can be introduced which requires the `useEffect` to be called again. Without having the safetynet of `exhaustive-deps`, it is likely that this new dependency will be forgotten to be added. This is not a made up issue, this is something I've seen happen several times. This is also why I'm working on removing all the ignore statements for exhaustive-deps.

Reply:
Can you please show me the code you believe to be correct? I have shown my attempt previously in this thread, but it requires an extra `useRef` (hidden in `usePrevious`), which I don't think adds any clarity or maintainability. Do you think my attempt is wrong in any way? As you see it includes `props.handleDataChange`, so that is not the big issue.

I'm tired of you not understanding that this effect is only intended to run once after the options have been loaded, in order to set the default choice (preselectedOptionIndex). When that option is set, the user should be allowed to deselect the option and pick another option (or leave it empty) if they want to.

PS: now that I think about it, the `preselectedOptionIndex` does not make any sense to use in `checkbox` anyway. For radio buttons it makes total sense, but with checkboxes it should be allowed to leave every box empty, and with the current implementation the preselected index will be checked again if you load the page with an empty selection.